Dict: Added support for collections.defaultdict

architv opened the pull request:
This commit should fix #10433 
- Added support for `collections.defaultdict` which was previously not
  available
- Used `isinstance` for comparing class types instead of direct class
  comparison

Example:

Previously sympy didn't support `collections.defaultdict`:

```
In [1]: from sympy import Dict

In [2]: from collections import defaultdict

In [3]: d = defaultdict(int)

In [4]: d[0]
Out[4]: 0

In [5]: Dict(d)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-3ed73043b7a9> in <module>()
----> 1 Dict(d)

/usr/local/lib/python2.7/dist-packages/sympy/core/containers.pyc in __new__(cls, *args)
    201             items = [Tuple(k, v) for k, v in args]
    202         else:
--> 203             raise TypeError('Pass Dict args as Dict((k1, v1), ...) or Dict({k1: v1, ...})')
    204         elements = frozenset(items)
    205         obj = Basic.__new__(cls, elements)

TypeError: Pass Dict args as Dict((k1, v1), ...) or Dict({k1: v1, ...})
```

Now:

```
In [1]: from sympy import Dict

In [2]: from collections import defaultdict

In [3]: d = defaultdict(int)

In [4]: d[0]
Out[4]: 0

In [5]: Dict(d)
Out[5]: {0: 0}

In [6]: a = Dict(d)

In [7]: a.items()
Out[7]: [(0, 0)]
```

architv committed:
Dict: Added support for collections.defaultdict

- Added support for collections.defaultdict which was previously not
 available
- Used isinstance for comparing class types instead of direct class
 comparison

Example:

In [1]: from sympy import Dict

In [2]: from collections import defaultdict

In [3]: d = defaultdict(int)

In [4]: d[0]
Out[4]: 0

In [5]: Dict(d)
Out[5]: {0: 0}

In [6]: a = Dict(d)

In [7]: a.items()
Out[7]: [(0, 0)]

AnishShah commented:
@architv (disclosure: not dev) you might want to add some tests?

architv commented:
@AnishShah I am not much familiar with writing tests. What all tests should I write?

AnishShah commented:
@architv In the first comment, you wrote a code snippet (previous/now). Instead you need to write tests to prove that your changes are working. Go to `sympy/core/tests/test_containers.py` and add tests inside `test_Dict()` 

Then execute the tests to check if they are working. [link] (https://github.com/sympy/sympy#tests)

AnishShah commented on sympy/core/containers.py:
```diff
@@ -194,8 +195,9 @@ class Dict(Basic):
-                             (args[0].__class__ is Dict)):
+        if len(args) == 1 and (isinstance(args[0], dict) or
+            isinstance(args[0], Dict) or
+            isinstance(args[0], defaultdict)):
```
I think, you can just use `args[0].__class__ is defaultdict` instead of using isinstance.

jksuom replied:
Is that condition necessary? An instance of ``defaultdict`` is also an instance of ``dict``.

AnishShah replied:
> Is that condition necessary? An instance of defaultdict is also an
instance of dict.
>
> —

defaultdict is inherited from dict. But, `__class__` returns the type which
is `defaultdict`. So, it won't work.

architv replied:
@jksuom You are right, `defaultdict` is an instance `dict`. So, checking the `isinstance` condition for `defaultdict` isn't of any use. I'll remove that.

@AnishShah Direct class comparison can be done but it would be better to use `isinstance`. See @asmeurer's comment in  #10433.

architv committed:
Dict: Removed unnecessary check for defaultdict

Checking the isinstance condition for defaultdict wasn't of any
use as defaultdict is also an instance of dict.

architv committed:
Dict: Revised tests and removed import of defaultdict in
sympy/core/containers.py

- Added tests to check the support for defaultdict with sympy's
Dict.
- Removed the import of collections.defaultdict from
sympy/core/containers.py as it is no longer required.

architv commented:
@AnishShah I have added some tests. Should I add more? Please let me know.

asmeurer commented on sympy/core/containers.py:
```diff
@@ -194,8 +194,8 @@ class Dict(Basic):
     def __new__(cls, *args):
-        if len(args) == 1 and ((args[0].__class__ is dict) or
-                             (args[0].__class__ is Dict)):
+        if len(args) == 1 and (isinstance(args[0], dict) or
```
You can use a single isinstance call here `isinstance(args[0], (dict, Dict))`.

asmeurer commented:
Looks good.

architv committed:
Changed the if condition to use single isinstance call

architv commented:
@asmeurer I have changed the condition to use a single `isinstance` call.

asmeurer commented:
+1 to merge as soon as the tests come in.

architv commented:
@asmeurer I am not able to figure out why the tests failed when I changed the `if` condition to use the single `isinstance` call. Should I revert to using two calls?

jksuom commented:
It looks like an unrelated PyPy failure. I've restarted the job.

jksuom commented:
Merging.
